File: ROS/offboard/src/pbvs_node.cpp
#include "utils.h"
#include "vision_utils.h"
#include <Eigen/Dense>

#include <ros/ros.h>
#include "geometry_msgs/Twist.h"
#include "tf/transform_broadcaster.h"
#include <image_transport/image_transport.h>
#include <opencv2/highgui/highgui.hpp>
#include <cv_bridge/cv_bridge.h>
#include <iostream>
#include <stdlib.h>

#include <std_msgs/Float64MultiArray.h>
#include <std_msgs/Float64.h>
#include <eigen_conversions/eigen_msg.h>

cv::Mat refImage, currentImage;
double t0,t;

static const std::string REF_IMAGE_NAME = "/home/noe/catkin_ws/src/offboard/image_reference/reference_image.jpg";

void imageCallback(const sensor_msgs::ImageConstPtr& msg)
{

  cv_bridge::CvImagePtr cvPtr;

	try
	{ 
		cvPtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
	}
	catch (cv_bridge::Exception& e) 
	{
		ROS_ERROR("cv_bridge exception: %s", e.what());
		return;
	}
  cvPtr->image.copyTo(currentImage);
}



int main(int argc, char **argv)
{
    Matrix_t K(3,3); //Calibration of the camera matrix
    Matrix_t H(3,3); // Homography matrix
    Matrix_t R(3,3); // Rotation between current frame and the reference frame
    Vector3D_t t;    // Translation between current frame and the reference frame
    Vector3D_t n;    // normal vector in the reference plane
    Vector3D_t u;   // Rodrigues vector
    Vector3D_t Uv;   // Control input for the translational velocity
    Vector3D_t Uw;   // Control input for the rotational velocity

    FPTYPE d;       // Distance between the camera and the reference plane
    homographySolution homSolution;
    int counter = 0;


    K << 554.382713,     0.0   , 320.0, 
         0.0       , 554.382713, 240.0, 
         0.0       ,     0.0   ,   1.0;
    if (argc < 4) {
        std::cerr << "Usage: " << argv[0] << " <topic_name> <lambdav> <lambdaw>" << std::endl;
        return 1;
    }

    
    FPTYPE lambdav = atof(argv[2]);
    FPTYPE lambdaw = atof(argv[3]);

    refImage = cv::imread(REF_IMAGE_NAME, cv::IMREAD_COLOR);
    if(refImage.empty())
    {
        std::cout << "Could not read the image: " << REF_IMAGE_NAME << std::endl;
        return 1;
    }
    else{
        std::cout << "Ref image is ready: " << REF_IMAGE_NAME << std::endl;
    }

    ros::init(argc, argv, "pbvs_node");
    ros::NodeHandle nh;

    cv::startWindowThread();
    image_transport::ImageTransport it(nh);
    
    std::string topicName;
    if(argc != 1){
      topicName = argv[1];
    } 
    else
    {
      topicName = "/my_image";
      std::cout <<"No topic name given. The default topic name will be used: " << topicName << std::endl;
    }

    ros::Publisher homography_pub   = nh.advertise<std_msgs::Float64MultiArray>("/homograpy_numerical",10);
    ros::Publisher ck_t_ct_pub      = nh.advertise<std_msgs::Float64MultiArray>("/ck_t_ct",10);
    ros::Publisher n_pub            = nh.advertise<std_msgs::Float64MultiArray>("/n_plane_vector",10);
    ros::Publisher Uv_pub           = nh.advertise<std_msgs::Float64MultiArray>("/Uv",10);
    ros::Publisher Uw_pub           = nh.advertise<std_msgs::Float64MultiArray>("/Uw",10);
    ros::Publisher d_pub            = nh.advertise<std_msgs::Float64>("/d_value",10);


    image_transport::Subscriber sub = it.subscribe(topicName, 1, imageCallback);

    //ros::spin();
    ros::Rate loop_rate(10); //Se ejecuta 100 por segundo
    ros::Duration(2).sleep(); //Retardo  de tiempo

    std_msgs::Float64MultiArray H_msg,ck_t_ct_msg,n_msg, Uv_msg, Uw_msg;
    std_msgs::Float64 d_msg;

    do{
          if(currentImage.empty())
          {
              std::cout << "Could not read the current image: "  << std::endl;
          }
          else
          {
              //std::cout << "current image is ready: "  << std::endl;
              cv::imwrite("/home/noe/catkin_ws/src/offboard/src/current_image.jpg",currentImage);
              EstimateHomography(refImage, currentImage, K, H, counter );
              RecoverFromHomography(H,R,t,n,d, counter, homSolution);
              Rodriguez(R,u);
              PBVSController(R,t,u,Uv,Uw,lambdav,lambdaw);
              
              tf::matrixEigenToMsg(Uv, Uv_msg);
              tf::matrixEigenToMsg(Uw, Uw_msg);
              tf::matrixEigenToMsg(H, H_msg);
              tf::matrixEigenToMsg(t, ck_t_ct_msg);
              tf::matrixEigenToMsg(n, n_msg);
              homography_pub.publish(H_msg);
              ck_t_ct_pub.publish(ck_t_ct_msg);
              n_pub.publish(n_msg);
              d_msg.data = d;
              d_pub.publish(d_msg);
              Uv_pub.publish(Uv_msg);
              Uw_pub.publish(Uw_msg);
          }
          counter++;
          ros::spinOnce();   //Activa callbacks
          loop_rate.sleep(); //Espera hasta completar el loop rate
    }while(ros::ok());

    //cv::destroyWindow("Video Stream");
    return 0;
}

File: ROS2/offboard/ros2_offboard/pbvs_node.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from std_msgs.msg import Float64MultiArray, Float64
from cv_bridge import CvBridge, CvBridgeError
import numpy as np
import cv2

from .vision_utils import estimate_homography, recover_from_homography, rodriguez, pbvs_controller

REF_IMAGE_NAME = '/home/noe/catkin_ws/src/offboard/image_reference/reference_image.jpg'

class PBVSNode(Node):
    def __init__(self):
        super().__init__('pbvs_node')
        self.declare_parameter('reference_image', REF_IMAGE_NAME)
        self.ref_image = cv2.imread(
            self.get_parameter('reference_image').get_parameter_value().string_value,
            cv2.IMREAD_COLOR)
        if self.ref_image is None:
            self.get_logger().error('Could not load reference image')
            raise RuntimeError('Failed to load reference image')
        self.current_image = None
        self.bridge = CvBridge()

        self.homography_pub = self.create_publisher(Float64MultiArray, '/homograpy_numerical', 10)
        self.ck_t_ct_pub = self.create_publisher(Float64MultiArray, '/ck_t_ct', 10)
        self.n_pub = self.create_publisher(Float64MultiArray, '/n_plane_vector', 10)
        self.Uv_pub = self.create_publisher(Float64MultiArray, '/Uv', 10)
        self.Uw_pub = self.create_publisher(Float64MultiArray, '/Uw', 10)
        self.d_pub = self.create_publisher(Float64, '/d_value', 10)

        self.sub = self.create_subscription(Image, '/iris/usb_cam/image_raw', self.image_callback, 10)

        self.K = np.array([[554.382713, 0.0, 320.0],
                           [0.0, 554.382713, 240.0],
                           [0.0, 0.0, 1.0]])
        self.counter = 0
        self.homography_solution = None
        self.timer = self.create_timer(0.1, self.timer_callback)

    def image_callback(self, msg):
        try:
            cv_image = self.bridge.imgmsg_to_cv2(msg, 'bgr8')
            self.current_image = cv_image
        except CvBridgeError as e:
            self.get_logger().error(f'CvBridge Error: {e}')

    def timer_callback(self):
        if self.current_image is None:
            return
        H = estimate_homography(self.ref_image, self.current_image, self.K, self.counter)
        if H is not None:
            R, t, n, d, self.homography_solution = recover_from_homography(
                H, self.K, self.counter, self.homography_solution)
            u = rodriguez(R)
            Uv, Uw = pbvs_controller(R, t, u, lambdav=0.5, lambdaw=0.5)

            self.homography_pub.publish(Float64MultiArray(data=H.flatten()))
            self.ck_t_ct_pub.publish(Float64MultiArray(data=t.flatten()))
            self.n_pub.publish(Float64MultiArray(data=n.flatten()))
            self.d_pub.publish(Float64(data=d))
            self.Uv_pub.publish(Float64MultiArray(data=Uv.flatten()))
            self.Uw_pub.publish(Float64MultiArray(data=Uw.flatten()))
        self.counter += 1


def main(args=None):
    rclpy.init(args=args)
    node = PBVSNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
